Add architecture specification

AG6GR opened the pull request:
This is a writeup of the architecture specification we drafted during 1/14 meeting. Everyone should read this and comment!

AG6GR committed:
Create ArchitectureSpec.md

Add architecture spec document based on team meeting notes.

AG6GR committed:
Update README.md with agreed branching strategy

AG6GR committed:
Add gitignore

AG6GR committed:
Add style checking Github action

AG6GR committed:
Permit lines longer than 79 characters in style check

Modify pycodestyle invocation to disable rule E501.

siddarth-pm commented:
Sounds good... Detection team specifications match up with what we're doing and as far as I know it matches up for others as well.

siddarth-pm reviewed (approved):
Sounds good... Detection team specifications match up with what we're doing and as far as I know it matches up for others as well.

michelllepan commented on ArchitectureSpec.md:
```diff
@@ -0,0 +1,74 @@
+Constructor takes in a *Detector object
+
+## API
+### run_cv()
```
Could we change the name of this method to `get_values()` to make it more readable?

AG6GR replied:
Sure!

AG6GR replied:
Will post a change once we get everyone's feedback.

michelllepan commented on ArchitectureSpec.md:
```diff
@@ -0,0 +1,74 @@
+### run_cv()
+Call Detector.run_detection()
+
+Return: `tuple(Euler Rotation: (yaw, pitch, roll) in degrees, World Position (x, y, z) in meters)`
```
We need to also provide another angle value for the relative heading of the target.

AG6GR replied:
Hm, mathematically there's only one set of translation and rotation. However you're right that we haven't defined what the reference point is.

We could define the camera position to be at the origin and the rotation to be zero. Then we would be returning the relative position of the target and the rotation of the target.

Or we could define the target to be at the origin and its rotation to be zero. Then we would be returning the relative position of the camera and the rotation of the camera.

PranavEranki replied:
Would defining the camera position to be at the origin be more consistent across all use cases? That way it may be easier for us to compare values when testing. However, is there any difference in calculations necessary between the two possible reference points, when it comes to the challenge?

AG6GR replied:
It's basically just a sign change. Or matrix inverse if you're thinking in terms of transform matrices.

PranavEranki commented on ArchitectureSpec.md:
```diff
@@ -0,0 +1,74 @@
+- `ImageGenerator (single image frame filename)`
+- `VideoFileGenerator(filename)`
+- `VideoLiveGenerator(camera port input int)`
+## API
```
Do any of the individual ImageGenerator subclasses have any additional functionality, or does this API match up across all subclasses?

AG6GR replied:
Same API across all subclasses. This enables the upper layers to use them interchangeably. The fancy name for this is polymorphism.

amoghs1912 commented on ArchitectureSpec.md:
```diff
@@ -0,0 +1,74 @@
+Gets data for one frame
+
+Returns: `tuple(rgb ndarray shape (x, y, 3), depth ndarray shape(x,y))`
+For sources with no depth data, return `tuple(ndarray, None)`
```
This section should mention that the numpy arrays `generate()` returns have the y-value first and then the x-value--(y, x, rgb)--since that's how opencv reads and uses arrays to represent images and depth data.

AG6GR replied:
Does this also apply to the depth array? Maybe for consistency we should return that in (row, column) order as well. It's a simple .transpose() call if it isn't already that way.

amoghs1912 replied:
The depth array is currently the same, (y, x).

AG6GR committed:
Address review comments

- Rename situational awareness API name.
- Define situational awareness coordinate system.
- Revise input image dimension order.

rithue commented on ArchitectureSpec.md:
```diff
@@ -0,0 +1,74 @@
+- `VideoLiveGenerator(camera port input int)`
+## API
+### get_horizontal_fov()
+Returns float degrees
```
Can we change the return type to radians for the fov methods so the other teams can use it to calculate the focal length?

AG6GR replied:
Sure, but we should probably change all our angle units to radians in that case. @michelle-pan @gargideshpande any thoughts?

michelllepan replied:
I think the rotation angles would be better in degrees because the range of values is more intuitive. If we want all the angle units to be uniform, could we keep the FoV in degrees and just convert it if necessary?

rithue replied:
Yes, we can just keep it in degrees then.
